File: backend/src/services/chart_service.py
"""
Chart calculation service - orchestrates all astronomical calculations.
"""

from datetime import datetime
from typing import Optional

from ..core.ephemeris import get_planet_positions, SiderealCalculator, datetime_to_jd
from ..core.rashi import RASHIS, get_rashi
from ..core.nakshatra import get_nakshatra, NAKSHATRAS
from ..core.dasha import VimshottariDasha
from ..models.schemas import (
    BirthData,
    PlanetPositionResponse,
    NakshatraInfo,
    DashaPeriod,
    AntardashaPeriod,
    PratyantardashaPeriod,
    CurrentDashaResponse,
    DashaWithAntardashas,
    DashaTimelineResponse,
    ChartResponse,
    TransitResponse,
)


class ChartService:
    """Service for calculating Vedic astrology charts."""
    
    def __init__(self):
        pass
    
    def _position_to_response(
        self, 
        planet_name: str, 
        position
    ) -> PlanetPositionResponse:
        """Convert internal position to API response model."""
        nakshatra_info = get_nakshatra(position.longitude)
        
        return PlanetPositionResponse(
            planet=planet_name,
            longitude=round(position.longitude, 6),
            latitude=round(position.latitude, 6),
            rashi=RASHIS[position.rashi],
            rashi_index=position.rashi,
            rashi_degree=round(position.rashi_degree, 4),
            nakshatra=nakshatra_info['name'],
            nakshatra_index=position.nakshatra,
            nakshatra_pada=position.nakshatra_pada,
            is_retrograde=position.is_retrograde,
            speed=round(position.speed, 6)
        )
    
    def calculate_planet_positions(
        self, 
        birth_data: BirthData
    ) -> tuple[list[PlanetPositionResponse], PlanetPositionResponse]:
        """
        Calculate all planet positions for birth data.
        
        Returns:
            Tuple of (planet_positions, ascendant)
        """
        positions = get_planet_positions(
            dt=birth_data.date,
            latitude=birth_data.latitude,
            longitude=birth_data.longitude,
            timezone=birth_data.timezone,
            ayanamsa=birth_data.ayanamsa
        )
        
        planet_responses = []
        ascendant = None
        
        for planet_name, position in positions.items():
            response = self._position_to_response(planet_name, position)
            
            if planet_name == 'ASCENDANT':
                ascendant = response
            else:
                planet_responses.append(response)
        
        return planet_responses, ascendant
    
    def get_moon_nakshatra(self, birth_data: BirthData) -> NakshatraInfo:
        """Get Moon's nakshatra (Janma Nakshatra)."""
        positions = get_planet_positions(
            dt=birth_data.date,
            latitude=birth_data.latitude,
            longitude=birth_data.longitude,
            timezone=birth_data.timezone,
            ayanamsa=birth_data.ayanamsa
        )
        
        moon_position = positions['MOON']
        nakshatra_info = get_nakshatra(moon_position.longitude)
        
        return NakshatraInfo(
            index=nakshatra_info['index'],
            name=nakshatra_info['name'],
            lord=nakshatra_info['lord'],
            pada=nakshatra_info['pada'],
            degree=round(nakshatra_info['degree'], 4),
            deity=nakshatra_info['deity'],
            symbol=nakshatra_info['symbol'],
            gana=nakshatra_info['gana']
        )
    
    def _create_dasha_calculator(self, birth_data: BirthData) -> VimshottariDasha:
        """Create dasha calculator for birth data."""
        positions = get_planet_positions(
            dt=birth_data.date,
            latitude=birth_data.latitude,
            longitude=birth_data.longitude,
            timezone=birth_data.timezone,
            ayanamsa=birth_data.ayanamsa
        )
        
        moon_longitude = positions['MOON'].longitude
        return VimshottariDasha(moon_longitude, birth_data.date)
    
    def get_dasha_timeline(
        self, 
        birth_data: BirthData,
        years_ahead: int = 120
    ) -> DashaTimelineResponse:
        """
        Get complete Mahadasha/Antardasha timeline.
        
        Args:
            birth_data: Birth information
            years_ahead: How many years to generate
            
        Returns:
            Complete dasha timeline response
        """
        dasha_calc = self._create_dasha_calculator(birth_data)
        moon_nakshatra = self.get_moon_nakshatra(birth_data)
        
        # Get balance info
        balance = dasha_calc.calculate_dasha_balance()
        
        # Generate full timeline with antardashas
        full_timeline = dasha_calc.get_full_timeline_with_antardashas(years_ahead)
        
        timeline_response = []
        for item in full_timeline:
            md = item['mahadasha']
            mahadasha = DashaPeriod(
                lord=md['lord'],
                start=md['start'],
                end=md['end'],
                duration_years=md['years'],
                duration_days=md['years'] * 365.25,
                is_birth_dasha=md['is_birth_dasha']
            )
            
            antardashas = [
                AntardashaPeriod(
                    lord=ad['lord'],
                    start=ad['start'],
                    end=ad['end'],
                    duration_days=ad['days'],
                    mahadasha_lord=md['lord']
                )
                for ad in item['antardashas']
            ]
            
            timeline_response.append(DashaWithAntardashas(
                mahadasha=mahadasha,
                antardashas=antardashas
            ))
        
        return DashaTimelineResponse(
            birth_data=birth_data,
            moon_nakshatra=moon_nakshatra,
            birth_dasha_lord=balance['lord'],
            dasha_balance={
                'total_years': balance['total_years'],
                'elapsed_years': round(balance['elapsed_years'], 4),
                'remaining_years': round(balance['remaining_years'], 4),
                'remaining_days': round(balance['remaining_days'], 2)
            },
            timeline=timeline_response
        )
    
    def get_current_periods(
        self, 
        birth_data: BirthData,
        target_date: Optional[datetime] = None
    ) -> CurrentDashaResponse:
        """
        Get currently running Dasha/Antardasha/Pratyantardasha.
        
        Args:
            birth_data: Birth information
            target_date: Date to check (default: now)
            
        Returns:
            Current periods response
        """
        if target_date is None:
            target_date = datetime.now()
        
        dasha_calc = self._create_dasha_calculator(birth_data)
        current = dasha_calc.get_current_periods(target_date)
        
        if 'error' in current:
            raise ValueError(current['error'])
        
        mahadasha = DashaPeriod(
            lord=current['mahadasha']['lord'],
            start=current['mahadasha']['start'],
            end=current['mahadasha']['end'],
            duration_years=current['mahadasha']['years'],
            duration_days=current['mahadasha']['years'] * 365.25,
            is_birth_dasha=False
        )
        
        antardasha = AntardashaPeriod(
            lord=current['antardasha']['lord'],
            start=current['antardasha']['start'],
            end=current['antardasha']['end'],
            duration_days=current['antardasha']['days'],
            mahadasha_lord=current['mahadasha']['lord']
        )
        
        pratyantardasha = None
        if current.get('pratyantardasha'):
            pd = current['pratyantardasha']
            pratyantardasha = PratyantardashaPeriod(
                lord=pd['lord'],
                start=pd['start'],
                end=pd['end'],
                duration_days=pd['days'],
                mahadasha_lord=current['mahadasha']['lord'],
                antardasha_lord=current['antardasha']['lord']
            )
        
        return CurrentDashaResponse(
            target_date=target_date,
            mahadasha=mahadasha,
            antardasha=antardasha,
            pratyantardasha=pratyantardasha
        )
    
    def calculate_full_chart(self, birth_data: BirthData) -> ChartResponse:
        """
        Generate complete Vedic birth chart with Dasha.
        
        Args:
            birth_data: Birth information
            
        Returns:
            Complete chart response
        """
        # Calculate positions
        planets, ascendant = self.calculate_planet_positions(birth_data)
        
        # Get Moon nakshatra
        moon_nakshatra = self.get_moon_nakshatra(birth_data)
        
        # Get current dasha
        current_dasha = self.get_current_periods(birth_data)
        
        # Get mahadasha timeline (simplified, without antardashas)
        dasha_calc = self._create_dasha_calculator(birth_data)
        mahadashas = dasha_calc.generate_mahadasha_timeline(years_ahead=120)
        
        mahadasha_timeline = [
            DashaPeriod(
                lord=md.lord,
                start=md.start,
                end=md.end,
                duration_years=md.years,
                duration_days=md.days,
                is_birth_dasha=md.is_birth_dasha
            )
            for md in mahadashas
        ]
        
        # Get ayanamsa value
        calculator = SiderealCalculator(birth_data.ayanamsa)
        jd = datetime_to_jd(birth_data.date, birth_data.timezone)
        ayanamsa_value = calculator.get_ayanamsa(jd)
        
        return ChartResponse(
            birth_data=birth_data,
            ayanamsa_value=round(ayanamsa_value, 6),
            planets=planets,
            ascendant=ascendant,
            moon_nakshatra=moon_nakshatra,
            current_dasha=current_dasha,
            mahadasha_timeline=mahadasha_timeline
        )
    
    def calculate_transits(
        self, 
        birth_data: BirthData,
        transit_date: datetime
    ) -> TransitResponse:
        """
        Get planetary transits for a specific date.
        
        Args:
            birth_data: Birth information
            transit_date: Date to calculate transits
            
        Returns:
            Transit response with natal and transit positions
        """
        # Get natal positions
        natal_planets, _ = self.calculate_planet_positions(birth_data)
        
        # Create transit birth data (same location, different time)
        transit_birth = BirthData(
            date=transit_date,
            latitude=birth_data.latitude,
            longitude=birth_data.longitude,
            timezone=birth_data.timezone,
            ayanamsa=birth_data.ayanamsa
        )
        
        # Get transit positions
        transit_planets, _ = self.calculate_planet_positions(transit_birth)
        
        return TransitResponse(
            transit_date=transit_date,
            natal_positions=natal_planets,
            transit_positions=transit_planets
        )

File: frontend/src/lib/services/chartService.ts
/**
 * Chart Service — orchestrates all local calculations.
 * Ephemeris calls are async (Swiss Ephemeris WASM is lazy-loaded).
 */

import { getPlanetPositions, getAyanamsaValue, type PlanetPosition as EphemerisPlanetPosition } from '../core/ephemeris';
import { RASHIS } from '../core/rashi';
import { getNakshatra } from '../core/nakshatra';
import { VimshottariDasha } from '../core/dasha';
import { YogaCalculator, type YogaResult } from '../core/yogas';
import type {
  BirthData, Chart, DashaTimeline, CurrentDasha,
  PlanetPosition, NakshatraInfo, DashaPeriod, AntardashaPeriod,
  PratyantardashaPeriod, DashaWithAntardashas,
} from '../../types/astrology';

function toIso(d: Date): string { return d.toISOString(); }

function toFrontendPosition(name: string, pos: EphemerisPlanetPosition): PlanetPosition {
  const nak = getNakshatra(pos.longitude);
  return {
    planet: name,
    longitude: Math.round(pos.longitude * 1000000) / 1000000,
    latitude: Math.round(pos.latitude * 1000000) / 1000000,
    rashi: RASHIS[pos.rashi],
    rashiIndex: pos.rashi,
    rashiDegree: Math.round(pos.rashiDegree * 10000) / 10000,
    nakshatra: nak.name,
    nakshatraIndex: nak.index,
    nakshatraPada: nak.pada,
    isRetrograde: pos.isRetrograde,
    speed: Math.round(pos.speed * 1000000) / 1000000,
  };
}

function toNakshatraInfo(nakData: ReturnType<typeof getNakshatra>): NakshatraInfo {
  return {
    index: nakData.index,
    name: nakData.name,
    lord: nakData.lord,
    pada: nakData.pada,
    degree: Math.round(nakData.degree * 10000) / 10000,
    deity: nakData.deity,
    symbol: nakData.symbol,
    gana: nakData.gana,
  };
}

export class ChartService {

  async calculatePlanetPositions(bd: BirthData): Promise<{ planets: PlanetPosition[]; ascendant: PlanetPosition }> {
    const positions = await getPlanetPositions(bd.date, bd.latitude, bd.longitude, bd.timezone, bd.ayanamsa);
    const planets: PlanetPosition[] = [];
    let ascendant!: PlanetPosition;
    for (const [name, pos] of Object.entries(positions)) {
      const fp = toFrontendPosition(name, pos);
      if (name === 'ASCENDANT') ascendant = fp;
      else planets.push(fp);
    }
    return { planets, ascendant };
  }

  async getMoonNakshatra(bd: BirthData): Promise<NakshatraInfo> {
    const positions = await getPlanetPositions(bd.date, bd.latitude, bd.longitude, bd.timezone, bd.ayanamsa);
    return toNakshatraInfo(getNakshatra(positions['MOON'].longitude));
  }

  /** Calculate yogas from birth chart positions */
  async calculateYogas(bd: BirthData): Promise<YogaResult[]> {
    const rawPositions = await getPlanetPositions(bd.date, bd.latitude, bd.longitude, bd.timezone, bd.ayanamsa);
    const posMap: Record<string, number> = {};
    for (const [name, pos] of Object.entries(rawPositions)) {
      if (name !== 'ASCENDANT') posMap[name] = pos.rashi;
    }
    const ascendantRashi = rawPositions['ASCENDANT'].rashi;
    const calc = new YogaCalculator(posMap, ascendantRashi);
    return calc.detectAllYogas();
  }

  private async _createDashaCalc(bd: BirthData): Promise<VimshottariDasha> {
    const positions = await getPlanetPositions(bd.date, bd.latitude, bd.longitude, bd.timezone, bd.ayanamsa);
    return new VimshottariDasha(positions['MOON'].longitude, new Date(bd.date));
  }

  async getDashaTimeline(bd: BirthData, yearsAhead = 120): Promise<DashaTimeline> {
    const calc = await this._createDashaCalc(bd);
    const moonNak = await this.getMoonNakshatra(bd);
    const balance = calc.calculateDashaBalance();
    const fullTimeline = calc.getFullTimelineWithAntardashas(yearsAhead);

    const timeline: DashaWithAntardashas[] = fullTimeline.map(item => ({
      mahadasha: {
        lord: item.mahadasha.lord,
        start: toIso(item.mahadasha.start),
        end: toIso(item.mahadasha.end),
        durationYears: item.mahadasha.years,
        durationDays: item.mahadasha.years * 365.25,
        isBirthDasha: item.mahadasha.isBirthDasha,
      },
      antardashas: item.antardashas.map(ad => ({
        lord: ad.lord,
        start: toIso(ad.start),
        end: toIso(ad.end),
        durationDays: ad.days,
        mahadashaLord: item.mahadasha.lord,
      })),
    }));

    return {
      birthData: bd,
      moonNakshatra: moonNak,
      birthDashaLord: balance.lord,
      dashaBalance: {
        totalYears: balance.totalYears,
        elapsedYears: Math.round(balance.elapsedYears * 10000) / 10000,
        remainingYears: Math.round(balance.remainingYears * 10000) / 10000,
        remainingDays: Math.round(balance.remainingDays * 100) / 100,
      },
      timeline,
    };
  }

  async getCurrentPeriods(bd: BirthData, targetDate?: Date): Promise<CurrentDasha> {
    const td = targetDate ?? new Date();
    const calc = await this._createDashaCalc(bd);
    const result = calc.getCurrentPeriods(td);
    if ('error' in result) throw new Error(result.error);

    const mahadasha: DashaPeriod = {
      lord: result.mahadasha.lord,
      start: toIso(result.mahadasha.start),
      end: toIso(result.mahadasha.end),
      durationYears: result.mahadasha.years,
      durationDays: result.mahadasha.years * 365.25,
      isBirthDasha: false,
    };
    const antardasha: AntardashaPeriod = {
      lord: result.antardasha.lord,
      start: toIso(result.antardasha.start),
      end: toIso(result.antardasha.end),
      durationDays: result.antardasha.days,
      mahadashaLord: result.mahadasha.lord,
    };
    let pratyantardasha: PratyantardashaPeriod | undefined;
    if (result.pratyantardasha) {
      pratyantardasha = {
        lord: result.pratyantardasha.lord,
        start: toIso(result.pratyantardasha.start),
        end: toIso(result.pratyantardasha.end),
        durationDays: result.pratyantardasha.days,
        mahadashaLord: result.mahadasha.lord,
        antardashaLord: result.antardasha.lord,
      };
    }
    return { targetDate: toIso(td), mahadasha, antardasha, pratyantardasha };
  }

  async calculateFullChart(bd: BirthData): Promise<Chart> {
    const { planets, ascendant } = await this.calculatePlanetPositions(bd);
    const moonNakshatra = await this.getMoonNakshatra(bd);
    const currentDasha = await this.getCurrentPeriods(bd);
    const calc = await this._createDashaCalc(bd);
    const mahadashas = calc.generateMahadashaTimeline(120);
    const mahadashaTimeline: DashaPeriod[] = mahadashas.map(md => ({
      lord: md.lord,
      start: toIso(md.start),
      end: toIso(md.end),
      durationYears: md.years,
      durationDays: md.days,
      isBirthDasha: md.isBirthDasha,
    }));
    const rawAyanamsa = await getAyanamsaValue(bd.date, bd.timezone, bd.ayanamsa);
    const ayanamsaValue = Math.round(rawAyanamsa * 1000000) / 1000000;
    return { birthData: bd, ayanamsaValue, planets, ascendant, moonNakshatra, currentDasha, mahadashaTimeline };
  }
}

export const chartService = new ChartService();
